packages/react-native/ReactCommon/react/nativemodule/core/platform/ios/ReactCommon/RCTTurboModule.mm

```diff
@@ -359,13 +400,52 @@ id convertJSIValueToObjCObject(jsi::Runtime &runtime, const jsi::Value &value, s
           rejectWasCalled = YES;
         };
 
-        invokeCopy(resolveBlock, rejectBlock);
+        RCTPromiseRejectBlock rejectBlock = ^(NSString *code, NSString *message, NSError *error) {
+          NSDictionary *reason = RCTJSErrorFromCodeMessageAndNSError(code, message, error);
+          handlePromiseRejection(reason);
+        };
+
+        RCTNSDictionaryPromiseRejectBlock internalRejectBlock = nil;
+
+        if (RCTRejectTurboModulePromiseOnNativeError()) {
+          internalRejectBlock = ^(NSDictionary *dict) {
+            handlePromiseRejection(dict);
+          };
+        }
+
+        invokeCopy(resolveBlock, rejectBlock, internalRejectBlock);
         return jsi::Value::undefined();
       });
 
   return Promise.callAsConstructor(runtime, fn);
 }
 
+static std::optional<jsi::JSError> handleCause(
```

Comment:
This method seems to have an awful lot of different behaviours:
1) It can return an error object (if isSync is set)
2) It can call a reject block
3) It can throw
4) It can do nothing (what happens with the error then?)

Reply:
Yeah, it does, but the same exact code is required in every catch block.

2 and 4 are one behavior. When reject callback is called `nullopt` is returned.
3 is needed for backward compatibility in cases where rejection on the native throw is not enabled we have to rethrow the error.

Reply:
Maybe we can more the logic under the try catch instead.